packages/context/src/interception-proxy.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/context
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Context} from './context';
+import {InvocationArgs, invokeMethodWithInterceptors} from './interceptor';
+
+/**
+ * Create the Promise type for `T`. If `T` extends `Promise`, the type is `T`,
+ * otherwise the type is `Promise<T>`.
+ */
+export type AsPromise<T> = T extends Promise<unknown> ? T : Promise<T>;
+
+/**
+ * The async variant of a function to always return Promise<R>. If T is not a
+ * function, the type is `T`.
+ */
+// tslint:disable-next-line:no-unused (possible tslint bug to treat `R` as unused)
+export type AsAsyncFunction<T> = T extends (...args: InvocationArgs) => infer R
+  ? (...args: InvocationArgs) => AsPromise<R>
```

Comment:
Per my earlier comments, it may be better to use `ValueOrPromise` here, to preserve sync behavior when both the intercepted method and all interceptors are synchronous.

IMO, modifying this API from `Promise` to `ValueOrPromise` would be breaking change, therefore it's better to get it right on the first try.